### PR TITLE
rl: persist gradient momentum evidence for policy updates

### LINKED ISSUE #1355: P1: Add momentum accumulation for high-variance RL policy gradients
## Problem
The latest Loop A training execution ledger reports repeated `POLICY_GRADIENT_HIGH_VARIANCE` as a self-actionable P1 capability gap. The current REINFORCE v1 policy-gradient path is producing nonzero gradients, but every recent report marks `gradientStable=false` / `highVariance=true`, and `previousGradientPresent=false`, so each run computes an isolated gradient with no inter-run smoothing.

## Evidence
- Ledger: `runtime-artifacts/rl-control-loop/20260523T104501Z-training-ledger.json`
- Status: `RUN_WITH_ANOMALY`, `trainingDidRun=true`
- Anomaly: `POLICY_GRADIENT_HIGH_VARIANCE`, `handoffRequired=true`, `handoffSeverity=P1`, `blockerClass=capability-gap`
- Evidence paths named by Loop A:
  - `runtime-artifacts/tencent-cloud/batch-runs/tencent-pg-20260523t101444z/remote/runtime-artifacts/rl-training/tencent-pg-20260523t101444z.json`
  - `runtime-artifacts/tencent-cloud/batch-runs/tencent-pg-20260523t092504z/remote/runtime-artifacts/rl-training/tencent-pg-20260523t092504z.json`
  - `runtime-artifacts/tencent-cloud/batch-runs/tencent-pg-20260523t085004z/remote/runtime-artifacts/rl-training/tencent-pg-20260523t085004z.json`
- Prior gradient-stabilization work such as #1314/#1345 is closed/Done, but this latest ledger still reports the current recurrence.

## Scope
Implement the smallest construction slice that makes policy-gradient updates less noisy and auditable, for example:
- persist/load previous gradient evidence across runs;
- apply exponential moving average or other momentum accumulation;
- record `previousGradientPresent=true` when used;
- add gradient clipping or variance diagnostics if needed;
- add focused tests and keep all outputs offline/private/shadow (`liveEffect=false`, `officialMmoWrites=false`).

## Acceptance criteria
- A focused test proves previous-gradient/momentum state is consumed and recorded.
- A focused test proves safety flags remain offline/private and no official MMO writes are introduced.
- A later Loop A/Tencent report shows the new momentum evidence fields and either lower variance or an explicit remaining blocker.

Refs #879
Refs #1032
Refs #1337

### PR DESCRIPTION
**Pull request by lanyusea**
## Summary
- Persists policy-gradient momentum state artifacts keyed by gradient comparison identity so later runs can load same-scheme previous-gradient evidence.
- Adds auditable `previousGradientPresent`, `previousGradientUsed`, state path/hash, and trust-gate evidence-source metadata while preserving offline-only safety flags.
- Keeps high-variance/conflicting-gradient candidates untrusted; momentum evidence improves traceability, not promotion laxity.

Fixes #1355

## Verification
- `git diff --check -- scripts/screeps_rl_training_runner.py scripts/test_screeps_rl_training_runner.py`
- `python3 -m py_compile scripts/screeps_rl_training_runner.py scripts/test_screeps_rl_training_runner.py`
- `python3 -m unittest scripts/test_screeps_rl_training_runner.py` — 138 tests OK
- `git merge-base --is-ancestor origin/main HEAD`

## Issue closure gate
- [x] #1355: Commit `21a53d381d6206ebe9d7ac0788aca427ed734ba3` adds persisted gradient-momentum state loading/materialization plus focused tests proving same-comparison reload and conservative untrusted classification for conflicting gradients.

## Universal task Done gate
- [x] Task type: code.
- [x] Expected observable outcome / named deliverable: Gradient momentum state is materialized under the training output directory and reused by later same-comparison training reports with explicit state identity/path/hash metadata.
- [x] Non-goals / accepted substitutes: No live MMO policy control, no Tencent paid compute launch, no cron change, and no reward-weight decision change in this slice.
- [x] Verification evidence required before Done: The three verification commands above pass from the PR worktree and the branch is based on current `origin/main`.
- [x] Project Evidence / Next action / Blocked by: Project items for #1355 and this PR record commit `21a53d38`, verification commands, and the review/merge gate as the next action; no owner action is required.
- [x] Post-merge/deploy/runtime proof: Script-only offline RL training runner change; proof surface is merged main containing commit `21a53d38` plus successful issue-completion/prod-ci checks, with no official Screeps deploy required.
- [x] Named deliverable proof: Tests `test_gradient_momentum_state_persists_and_loads_same_comparison` and `test_reinforce_gradient_conflict_remains_untrusted_with_loaded_momentum` prove persisted state reuse and conservative safety behavior.

## Safety
- `liveEffect=false`, `officialMmoWrites=false`, and `officialMmoWritesAllowed=false` remain asserted in state/momentum artifacts and tests.
- No secrets, runtime artifacts, `prod/node_modules`, or generated local caches are committed.

### CODE REVIEW

**Comment by lanyusea** [comment]
### RL steward update — PR opened for #1355

Recovered the vanished Codex lane for gradient momentum and preserved Codex authorship.

- PR: #1619 (`fix/issue-1355-gradient-momentum`)
- Commit: `21a53d381d6206ebe9d7ac0788aca427ed734ba3` (`lanyusea's bot <lanyusea@gmail.com>`)
- Controller verification: `git diff --check`, `python3 -m py_compile scripts/screeps_rl_training_runner.py scripts/test_screeps_rl_training_runner.py`, and `python3 -m unittest scripts/test_screeps_rl_training_runner.py` (138 tests OK).
- Project `screeps`: #1355 and PR #1619 set to In review with Evidence/Next action/Blocked by refreshed.

Next action: re-query exact-head PR #1619 checks, CodeRabbit, review threads, and the >=15 minute gate; merge only when all gates pass. No owner action required.

**Comment by lanyusea** [comment]
### RL steward review-fix handoff — CodeRabbit top-level concerns

CodeRabbit completed PR #1619 review and posted assertive top-level correctness/runtime concerns. I started one bounded Codex review-fix/triage lane in the same PR worktree; no second independent RL lane was dispatched.

- Codex session: `proc_f9e92b6e3b06` (OS pid `3293116`)
- Worktree: `/root/screeps-worktrees/issue-1355-gradient-momentum`
- Prompt: `/tmp/pr1619-review-fix-prompt.txt`
- Review input: `/tmp/pr1619-coderabbit-comment.md`
- Exact prior head: `21a53d381d6206ebe9d7ac0788aca427ed734ba3`
- Required output: per-finding classifications; commit only if a critical fix is valid; no push/GitHub/Project/cron/runtime mutation.

ProjectV2 GraphQL is exhausted until `2026-06-02T16:59:05Z`, so field sync for this active review-fix handoff is recorded here via REST. Next steward/continuation slice must poll `proc_f9e92b6e3b06`, inspect the worktree, run controller verification, push any Codex-authored fix, and then re-query Project fields/review threads after GraphQL reset. No owner action required.

**Comment by lanyusea** [comment]
### RL steward review-fix update — #1355 / PR #1619

The vanished/stale Codex review-fix lane left useful changes. I verified, preserved Codex authorship with a commit-only recovery, and pushed the fix commit.

- New head: `c3538d28fdaefbecbcc6abf48421be34c9594786`
- Commit: `c3538d28 lanyusea's bot <lanyusea@gmail.com> rl: harden gradient momentum state review findings`
- Triage artifact: `/tmp/pr1619-codex-triage-20260602T1715Z.md` (side evidence only, not staged)
- Fixed review threads/classes:
  - `PRRT_kwDOSMSsO86Gf3SC`: secret scan before gradient-momentum state write.
  - `PRRT_kwDOSMSsO86Gf9CU`: payload-hash round-trip uses the same canonical hash.
  - `PRRT_kwDOSMSsO86Gf9CY`: prior state load error is preserved into report/state summary.
  - `PRRT_kwDOSMSsO86Gf9Cc`: persisted loaded momentum state is authoritative over bogus inline EMA values while preserving configured decay.
  - `PRRT_kwDOSMSsO86Gf9Ce`: gradient-momentum helper now builds a valid policy-gradient card.
  - `PRRT_kwDOSMSsO86Gf9Cp`: conflict test now exercises real state materialization + loader wiring.
- Controller verification after commit:
  - `git diff --check -- scripts/screeps_rl_training_runner.py scripts/test_screeps_rl_training_runner.py` PASS
  - `python3 -m py_compile scripts/screeps_rl_training_runner.py scripts/test_screeps_rl_training_runner.py` PASS
  - `python3 -m unittest scripts/test_screeps_rl_training_runner.py` PASS, 140 tests

Next gate: wait for exact-head GitHub checks and CodeRabbit/Gemini to finish on `c3538d28`; then re-query all review threads, resolve fixed/stale threads only after bot completion, refresh Project fields, and merge only if the same-run Screeps merge gate passes. No owner action required.

**Comment by lanyusea** [comment]
### RL steward review-fix redispatch — #1355 / PR #1619

Fresh gate re-query found PR #1619 is not merge-ready despite green checks: GraphQL currently has 3 unresolved non-outdated CodeRabbit threads on `c3538d28` plus one top-level/outside-diff comparison-key isolation concern. The previous dirty worktree already contained useful uncommitted fixes for the 3 inline findings but did not cover the outside-diff concern, and no Codex process was live for this worktree.

- Worktree: `/root/screeps-worktrees/issue-1355-gradient-momentum`
- New Codex session: `proc_01783a85dbfb` (OS pid `3359638`)
- Prompt: `/tmp/pr1619-unified-review-fix-prompt.txt`
- Required triage artifact: `/tmp/pr1619-codex-triage-20260603T0220Z.md`
- Scope: preserve/finalize dirty fixes for `PRRT_kwDOSMSsO86GhLWF`, `PRRT_kwDOSMSsO86GhLWN`, `PRRT_kwDOSMSsO86GhLWP`; triage/fix CodeRabbit outside-diff comparison-key isolation concern; run diff-check/py_compile/unittest; commit with Codex authorship; no push/GitHub/Project mutations from Codex.

Next action: next steward/continuation slice must poll `proc_01783a85dbfb`, verify the real worktree commit, push if valid, then re-query exact-head CodeRabbit/review threads/checks/mergeState and continue the normal PR gate. No owner action required.

**Comment by lanyusea** [comment]
### RL steward review-fix pushed — #1355 / PR #1619

Codex completed the unified review-fix lane and I verified/pushed the real worktree commit.

- New head: `9690881ae3fe8568844ada704edb0fd2293bdbf3`
- Commit: `9690881a lanyusea's bot <lanyusea@gmail.com> rl: harden gradient momentum state consistency`
- Triage artifact: `/tmp/pr1619-codex-triage-20260603T0220Z.md`
- Findings handled:
  - `PRRT_kwDOSMSsO86GhLWF`: FIX — rollback/delete materialized momentum state on final report secret-scan/write failure.
  - `PRRT_kwDOSMSsO86GhLWN`: FIX — require non-empty string `statePayloadHash`; malformed/missing hash records a load error.
  - `PRRT_kwDOSMSsO86GhLWP`: FIX — corrupt-state test now asserts the overwritten state artifact persists `previousStateArtifactLoadError`.
  - CodeRabbit outside-diff comparison-key isolation finding: FIX — added competing different-scheme state artifact in same output directory and asserts same-scheme state is loaded while competing artifact is ignored/unchanged.
- Controller verification after Codex commit:
  - `git diff --check -- scripts/screeps_rl_training_runner.py scripts/test_screeps_rl_training_runner.py` PASS
  - `python3 -m py_compile scripts/screeps_rl_training_runner.py scripts/test_screeps_rl_training_runner.py` PASS
  - `python3 -m unittest scripts/test_screeps_rl_training_runner.py` PASS — 142 tests
  - `git status --short --branch` clean

Current gate: exact-head GitHub checks and CodeRabbit are running/pending on `9690881a`; review threads are not resolved yet. Next steward/continuation slice should re-query exact-head checks, CodeRabbit, and all review threads; resolve fixed/stale threads only after the bot finishes, then merge only if the same-run Screeps merge gate passes. No owner action required.

**Comment by lanyusea** [comment]
### RL steward PR gate checkpoint — #1355 / PR #1619

Exact-head review-fix commit `9690881a` is pushed and locally/controller verified. GitHub checks are now green and GraphQL unresolved review threads are `[]`, but the PR is **not merge-ready** because the latest CodeRabbit top-level comment is a review-limit / usage-credit exhaustion notice rather than a substantive exact-head review. `mergeStateStatus` remains `BLOCKED` and `reviewDecision` remains stale `CHANGES_REQUESTED`.

Evidence:
- Exact head: `9690881ae3fe8568844ada704edb0fd2293bdbf3`
- Required checks: issue-completion-gate SUCCESS; prod-ci TypeScript/Jest/bundle SUCCESS; Pages contracts SUCCESS.
- Review threads: `[]` unresolved after GraphQL re-query.
- CodeRabbit status context: `SUCCESS`, but latest bot comment updated at 2026-06-02T18:29:08Z says review limit reached / usage credits exhausted; not a substantive review body.

Next action: do not merge or launch duplicate code work. After the review availability window, trigger/re-query CodeRabbit on exact head `9690881a`, then re-run the same-run merge gate (elapsed, checks, all threads, mergeState, Project fields). No owner action required.

**Comment by lanyusea** [comment]
@coderabbitai review

Requesting a fresh exact-head substantive review for PR #1619 at current head `9690881ae3fe8568844ada704edb0fd2293bdbf3`. Codex addressed the active momentum-state consistency findings and the controller verified/pushed the fix. Latest controller audit found zero unresolved GraphQL review threads and green required checks, but the PR remains merge-policy blocked by stale/previous CodeRabbit change-request state.

**Comment by lanyusea** [comment]
### RL steward gate checkpoint — #1355 / PR #1619

I re-audited the previously dispatched gradient-momentum lane.

- No Codex process is live for this worktree/session.
- Worktree: `/root/screeps-worktrees/issue-1355-gradient-momentum`, clean at `9690881ae3fe8568844ada704edb0fd2293bdbf3`.
- PR: #1619 remains open on `fix/issue-1355-gradient-momentum`.
- Controller verification re-run from the worktree: `git diff --check origin/main...HEAD` PASS; `python3 -m py_compile scripts/screeps_rl_training_runner.py scripts/test_screeps_rl_training_runner.py` PASS; `python3 -m unittest scripts/test_screeps_rl_training_runner.py` PASS (142 tests OK).
- REST exact-head checks: issue-completion gate SUCCESS, prod TypeScript/Jest/bundle SUCCESS, roadmap Pages contracts SUCCESS; CodeRabbit status context reports SUCCESS.
- REST review evidence: latest CodeRabbit substantive review is still the previous `CHANGES_REQUESTED` review on older heads; the later `@coderabbitai review` only produced a kick-off acknowledgement plus a review-limit/usage-credit notice, not a substantive exact-head no-blocker review.
- GitHub REST reports `mergeable_state=blocked`; GraphQL quota is exhausted until approximately 2026-06-02T20:00:34Z, so review-thread and ProjectV2 field verification are deferred.

Next action: do not merge and do not launch a duplicate code lane. After GraphQL quota resets and CodeRabbit availability returns, re-query exact-head review threads, Project fields, substantive CodeRabbit/Gemini evidence, elapsed gate, checks, and merge state; merge #1619 only if all gates clear. No owner action required.

**Comment by lanyusea** [comment]
@coderabbitai review

Fresh exact-head review request for PR #1619 at head `9690881ae3fe8568844ada704edb0fd2293bdbf3`. Controller re-verified diff-check, py_compile, and 142 unit tests; required checks are green and GraphQL unresolved review threads are empty. The remaining gate is stale CodeRabbit CHANGES_REQUESTED / mergeState BLOCKED, so please produce a substantive no-blocker review if the current head is clean.
